File: WordTarget/WordTargetCoreTests/WordTargetLayoutTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;

namespace WordTargetCore
{
    [TestClass]
    public class WordTargetLayoutTest
    {
        [TestMethod]
        public void LinkCreation()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE" });
            Assert.AreEqual(1, layout.GetWordsInCircle(5).Count);
            Assert.AreEqual("VARIABLE", layout.GetWordsInCircle(5)[0]);
            Assert.AreEqual(0, layout.GetWordsInCircle(4).Count);
            Assert.AreEqual(0, layout.GetWordsInCircle(3).Count);
            Assert.AreEqual(0, layout.GetWordsInCircle(2).Count);
            Assert.AreEqual("STATE", layout.WordInCenter);
        }

        // Assign word to a circle: error if it's not in the list or if it's first/last word;
        [TestMethod]
        public void AssignWord()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE" });
            layout.AssignWord("VALUE", 3);
            layout.AssignWord("UNCLEAR", 4);
            Assert.ThrowsException<System.Exception>(() => layout.AssignWord("NUCLEAR", 2));
            Assert.ThrowsException<System.Exception>(() => layout.AssignWord("VARIABLE", 3));
            Assert.ThrowsException<System.Exception>(() => layout.AssignWord("STATE", 1));
            Assert.ThrowsException<System.Exception>(() => layout.AssignWord("VAGUE", 1));
            Assert.ThrowsException<System.Exception>(() => layout.AssignWord("VALUE", 4));
            Assert.AreEqual("STATE", layout.WordInCenter);
            Assert.AreEqual(1, layout.GetWordsInCircle(3).Count);
            Assert.AreEqual(1, layout.GetWordsInCircle(4).Count);
            Assert.AreEqual("VALUE", layout.GetWordsInCircle(3)[0]);
            Assert.AreEqual("UNCLEAR", layout.GetWordsInCircle(4)[0]);


        }

        // Remove from other circles if there;
        [TestMethod]
        public void RemoveWord()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "UNCLEAR", "STATE" });
            layout.AssignWord("VALUE", 3);
            layout.AssignWord("UNCLEAR", 4);
            Assert.ThrowsException<System.Exception>(() => layout.RemoveWord("NUCLEAR", 2));
            Assert.ThrowsException<System.Exception>(() => layout.RemoveWord("UNCLEAR", 3));
            Assert.ThrowsException<System.Exception>(() => layout.RemoveWord("VARIABLE", 5));
            Assert.ThrowsException<System.Exception>(() => layout.RemoveWord("STATE", 1));
            layout.RemoveWord("VALUE", 3);
            Assert.AreEqual(0, layout.GetWordsInCircle(3).Count);
            Assert.AreEqual(1, layout.GetWordsInCircle(4).Count);
            Assert.AreEqual("UNCLEAR", layout.GetWordsInCircle(4)[0]);
        }

        // add to the target circle at the end of the list; calcualte the fraction and add it to the list of fractions for the circle
        [TestMethod]
        public void FracsWhenAssigningWords()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "UNCLEAR", "STATE" });
            Assert.AreEqual(0, layout.GetFracsInCircle(3).Count);
            layout.AssignWord("VALUE", 3);
            Assert.AreEqual(1, layout.GetFracsInCircle(3).Count);
            Assert.AreEqual(0.111, layout.GetFracsInCircle(3)[0], 0.001);
            layout.AssignWord("UNCLEAR", 3);
            Assert.AreEqual(2, layout.GetFracsInCircle(3).Count);
            Assert.AreEqual(0.162, layout.GetFracsInCircle(3)[1], 0.001);
        }

        [TestMethod]
        public void FracsWhenRemovingWords()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE" });
            layout.AssignWord("VALUE", 3);
            layout.AssignWord("UNCLEAR", 3);
            Assert.AreEqual(2, layout.GetFracsInCircle(3).Count);
            layout.RemoveWord("UNCLEAR", 3);
            Assert.AreEqual(1, layout.GetFracsInCircle(3).Count);
            layout.RemoveWord("VALUE", 3);
            Assert.AreEqual(0, layout.GetFracsInCircle(3).Count);
        }

        // Calculate empty space for circle: add up all the fractions, add min space for separator, return what's left
        [TestMethod]
        public void EmptySpace()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE" });
            layout.AssignWord("VALUE", 3);
            layout.AssignWord("UNCLEAR", 3);
            Assert.AreEqual(0.660, layout.GetEmptySpace(3), 0.001);
            Assert.AreEqual(1.0, layout.GetEmptySpace(4));
            Assert.AreEqual(0.892, layout.GetEmptySpace(5), 0.001);
            Assert.ThrowsException<System.Exception>(() => layout.GetEmptySpace(6));
            Assert.ThrowsException<System.Exception>(() => layout.GetEmptySpace(1));

        }

        // Is circle full? Check that there is no more space in the circle
        [TestMethod]
        public void IsCircleFull()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "STATE" });
            layout.AssignWord("VALUE", 3);
            layout.AssignWord("VAGUE", 3);
            layout.AssignWord("UNCLEAR", 3);
            layout.AssignWord("NUCLEAR", 3);
            Assert.AreEqual(false, layout.IsCircleFull(3));
            layout.AssignWord("CORE", 3);
            layout.AssignWord("CORN", 3);
            layout.AssignWord("COIN", 3);
            layout.AssignWord("JOIN", 3);
            Assert.ThrowsException<System.Exception>(() => layout.IsCircleFull(1));
            Assert.ThrowsException<System.Exception>(() => layout.IsCircleFull(6));
            Assert.AreEqual(true, layout.IsCircleFull(3));

        }

        // Where is the word? Return the circle or null if not assigned
        [TestMethod]
        public void FindWord()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE" });
            layout.AssignWord("VALUE", 3);
            layout.AssignWord("UNCLEAR", 4);
            layout.AssignWord("VAGUE", 2);
            Assert.AreEqual(2, layout.FindWord("VAGUE"));
            Assert.AreEqual(3, layout.FindWord("VALUE"));
            Assert.AreEqual(4, layout.FindWord("UNCLEAR"));
            Assert.AreEqual(5, layout.FindWord("VARIABLE"));
            Assert.AreEqual(1, layout.FindWord("STATE"));
            Assert.AreEqual(null, layout.FindWord("NUCLEAR"));
        }

        // Get previous and next words
        [TestMethod]
        public void GetNextWord()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE" });
            Assert.AreEqual("VALUE", layout.GetNextWord("VARIABLE"));
            Assert.AreEqual("VAGUE", layout.GetNextWord("VALUE"));
            Assert.AreEqual(null, layout.GetNextWord("STATE"));
            Assert.ThrowsException<System.Exception>(() => layout.GetNextWord("NOTTHERE"));
        }

        [TestMethod]
        public void GetPreviousWord()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE" });
            Assert.AreEqual(null, layout.GetPreviousWord("VARIABLE"));
            Assert.AreEqual("VARIABLE", layout.GetPreviousWord("VALUE"));
            Assert.AreEqual("UNCLEAR", layout.GetPreviousWord("STATE"));
            Assert.ThrowsException<System.Exception>(() => layout.GetNextWord("NOTTHERE"));
        }

        // Get words that are not assigned
        [TestMethod]
        public void GetUnassignedWords()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "STATE" });
            layout.AssignWord("VALUE", 2);
            layout.AssignWord("VAGUE", 2);
            List<string> unassigned = new List<string> { "UNCLEAR", "NUCLEAR" };
            Assert.IsTrue(Enumerable.SequenceEqual(unassigned, layout.GetUnassignedWords()));
        }

        [TestMethod]
        public void StartingAngles()
        {
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "STATE" });
            layout.SetStartingAngle(2, 80);
            layout.SetStartingAngle(3, 0);
            layout.SetStartingAngle(4, 5);
            Assert.AreEqual(80, layout.GetStartingAngle(2));
            Assert.AreEqual(0, layout.GetStartingAngle(3));
            Assert.AreEqual(5, layout.GetStartingAngle(4));
        }

        [TestMethod]
        public void DistributeRandomly()
        {
            Random rand = new Random(235);
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE" });
            layout.DistributeRandomly(rand);
            // TODO complete method
        }

        [TestMethod]
        public void EquilibrateOnce()
        {
            Random rand = new Random(236);
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE" });
            layout.DistributeRandomly(rand);
            layout.EquilibrateOnce(rand);
            // TODO complete method
        }

        [TestMethod]
        public void DoLayout()
        {
            Random rand = new Random(245);
            WordTargetLayout layout = new WordTargetLayout(new List<string> { "VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE" });
            layout.DoLayout(rand);
            // TODO complete method
            string wordTarget = Renderer.RenderWordTarget(layout);
            System.IO.File.WriteAllText(Environment.GetFolderPath(Environment.SpecialFolder.Desktop) + @"\WordTargetOutput.svg", wordTarget);
        }
    }
}

File: word-target/word-target-core/src/test/java/it/carcassi/wordtarget/core/WordTargetLayoutTest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package it.carcassi.wordtarget.core;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 *
 * @author Matteo
 */
public class WordTargetLayoutTest {

    @Test
    public void testLinkCreation() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE"));
        assertEquals(1, layout.getWordsInCircle(5).size());
        assertEquals("VARIABLE", layout.getWordsInCircle(5).get(0));
        assertEquals(0, layout.getWordsInCircle(4).size());
        assertEquals(0, layout.getWordsInCircle(3).size());
        assertEquals(0, layout.getWordsInCircle(2).size());
        assertEquals("STATE", layout.getWordInCenter());
    }

    // Assign word to a circle: error if it's not in the list or if it's first/last word;
    @Test
    public void testAssignWord() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE"));
        layout.assignWord("VALUE", 3);
        layout.assignWord("UNCLEAR", 4);
        assertThrows(IllegalArgumentException.class, () -> {
            layout.assignWord("NUCLEAR", 2);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.assignWord("VARIABLE", 3);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.assignWord("STATE", 1);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.assignWord("VAGUE", 1);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.assignWord("VALUE", 4);
        });
        assertEquals("STATE", layout.getWordInCenter());
        assertEquals(1, layout.getWordsInCircle(3).size());
        assertEquals(1, layout.getWordsInCircle(4).size());
        assertEquals("VALUE", layout.getWordsInCircle(3).get(0));
        assertEquals("UNCLEAR", layout.getWordsInCircle(4).get(0));

    }

    // Remove from other circles if there;
    @Test
    public void testRemoveWord() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "UNCLEAR", "STATE"));
        layout.assignWord("VALUE", 3);
        layout.assignWord("UNCLEAR", 4);
        assertThrows(IllegalArgumentException.class, () -> {
            layout.removeWord("NUCLEAR", 2);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.removeWord("UNCLEAR", 3);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.removeWord("VARIABLE", 5);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.removeWord("STATE", 1);
        });
        layout.removeWord("VALUE", 3);
        assertEquals(0, layout.getWordsInCircle(3).size());
        assertEquals(1, layout.getWordsInCircle(4).size());
        assertEquals("UNCLEAR", layout.getWordsInCircle(4).get(0));
    }

    // add to the target circle at the end of the list; calcualte the fraction and add it to the list of fractions for the circle
    @Test
    public void testFracsWhenAssigningWords() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "UNCLEAR", "STATE"));
        assertEquals(0, layout.getFracsInCircle(3).size());
        layout.assignWord("VALUE", 3);
        assertEquals(1, layout.getFracsInCircle(3).size());
        assertEquals(0.111, layout.getFracsInCircle(3).get(0), 0.001);
        layout.assignWord("UNCLEAR", 3);
        assertEquals(2, layout.getFracsInCircle(3).size());
        assertEquals(0.162, layout.getFracsInCircle(3).get(1), 0.001);
    }

    @Test
    public void testFracsWhenRemovingWords() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE"));
        layout.assignWord("VALUE", 3);
        layout.assignWord("UNCLEAR", 3);
        assertEquals(2, layout.getFracsInCircle(3).size());
        layout.removeWord("UNCLEAR", 3);
        assertEquals(1, layout.getFracsInCircle(3).size());
        layout.removeWord("VALUE", 3);
        assertEquals(0, layout.getFracsInCircle(3).size());
    }

    // Calculate empty space for circle: add up all the fractions, add min space for separator, return what's left
    @Test
    public void testEmptySpace() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE"));
        layout.assignWord("VALUE", 3);
        layout.assignWord("UNCLEAR", 3);
        assertEquals(0.660, layout.getEmptySpace(3), 0.001);
        assertEquals(1.0, layout.getEmptySpace(4));
        assertEquals(0.892, layout.getEmptySpace(5), 0.001);
        assertThrows(IllegalArgumentException.class, () -> {
            layout.getEmptySpace(6);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.getEmptySpace(1);
        });
    }

    // Is circle full? Check that there is no more space in the circle
    @Test
    public void testIsCircleFull() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "STATE"));
        layout.assignWord("VALUE", 3);
        layout.assignWord("VAGUE", 3);
        layout.assignWord("UNCLEAR", 3);
        layout.assignWord("NUCLEAR", 3);
        assertEquals(false, layout.isCircleFull(3));
        layout.assignWord("CORE", 3);
        layout.assignWord("CORN", 3);
        layout.assignWord("COIN", 3);
        layout.assignWord("JOIN", 3);
        assertThrows(IllegalArgumentException.class, () -> {
            layout.isCircleFull(1);
        });
        assertThrows(IllegalArgumentException.class, () -> {
            layout.isCircleFull(6);
        });
        assertEquals(true, layout.isCircleFull(3));

    }

    // Where is the word? Return the circle or null if not assigned
    @Test
    public void testFindWord() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE"));
        layout.assignWord("VALUE", 3);
        layout.assignWord("UNCLEAR", 4);
        layout.assignWord("VAGUE", 2);
        assertEquals(2, layout.findWord("VAGUE"));
        assertEquals(3, layout.findWord("VALUE"));
        assertEquals(4, layout.findWord("UNCLEAR"));
        assertEquals(5, layout.findWord("VARIABLE"));
        assertEquals(1, layout.findWord("STATE"));
        assertEquals(null, layout.findWord("NUCLEAR"));
    }

    // Get previous and next words
    @Test
    public void testGetNextWord() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE"));
        assertEquals("VALUE", layout.getNextWord("VARIABLE"));
        assertEquals("VAGUE", layout.getNextWord("VALUE"));
        assertEquals(null, layout.getNextWord("STATE"));
        assertThrows(IllegalArgumentException.class, () -> {
            layout.getNextWord("NOTTHERE");
        });
    }

    @Test
    public void testGetPreviousWord() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "STATE"));
        assertEquals(null, layout.getPreviousWord("VARIABLE"));
        assertEquals("VARIABLE", layout.getPreviousWord("VALUE"));
        assertEquals("UNCLEAR", layout.getPreviousWord("STATE"));
        assertThrows(IllegalArgumentException.class, () -> {
            layout.getNextWord("NOTTHERE");
        });
    }

    // Get words that are not assigned
    @Test
    public void testGetUnassignedWords() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "STATE"));
        layout.assignWord("VALUE", 2);
        layout.assignWord("VAGUE", 2);
        Set<String> unassigned = new HashSet<>(Set.of("NUCLEAR", "UNCLEAR"));
        assertEquals(unassigned, layout.getUnassignedWords());
    }

    @Test
    public void testStartingAngles() {
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "STATE"));
        layout.setStartingAngle(2, 80);
        layout.setStartingAngle(3, 0);
        layout.setStartingAngle(4, 5);
        assertEquals(80, layout.getStartingAngle(2));
        assertEquals(0, layout.getStartingAngle(3));
        assertEquals(5, layout.getStartingAngle(4));
    }

    @Test
    public void testDistributeRandomly() {
        Random rand = new Random(235);
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE"));
        layout.distributeRandomly(rand);
        // TODO complete method
    }

    @Test
    public void testEquilibrateOnce() {
        Random rand = new Random(236);
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE"));
        layout.distributeRandomly(rand);
        layout.equilibrateOnce(rand);
        // TODO complete method
    }

    @Test
    public void testDoLayout() {
        Random rand = new Random(249);
        WordTargetLayout layout = new WordTargetLayout(List.of("VARIABLE", "VALUE", "VAGUE", "UNCLEAR", "NUCLEAR", "CORE", "CORN", "COIN", "JOIN", "JOINT", "ANKLE", "ANGLE", "RIGHT", "SIGHT", "VITAMIN A", "CAROTENOIDS", "COORDINATES", "POLAR", "POLARS", "POLARIS", "STAR", "STARE", "SPARE", "SPACE", "STATION", "STATIN", "STATEN", "STATE"));
        layout.doLayout(rand);
        // TODO complete method
        String wordTarget = Renderer.renderWordTarget(layout);
        //System.IO.File.WriteAllText(Environment.GetFolderPath(Environment.SpecialFolder.Desktop) + @"\WordTargetOutput.svg", wordTarget);
    }
}
